substrate-archive/src/actors.rs

```diff
@@ -157,15 +144,79 @@ where
 	}
 }
 
-struct Actors<B: BlockT + Unpin, D: ReadOnlyDb + 'static>
+struct Actors<Block: Send + Sync + 'static, H: Send + Sync + 'static, Db: Send + Sync + 'static> {
+	storage: Address<workers::StorageAggregator<H>>,
+	blocks: Address<workers::BlocksIndexer<Block, Db>>,
+	metadata: Address<workers::MetadataActor<Block>>,
+	db: Address<DatabaseActor>,
+	tasks: Vec<Task<()>>,
+}
+
+impl<B: Send + Sync + 'static, H: Send + Sync + 'static, D: Send + Sync + 'static> Clone for Actors<B, H, D> {
+	fn clone(&self) -> Self {
+		Self {
+			storage: self.storage.clone(),
+			blocks: self.blocks.clone(),
+			metadata: self.metadata.clone(),
+			db: self.db.clone(),
+			tasks: Vec::new(),
+		}
+	}
+}
+
+impl<Block, Db> Actors<Block, Block::Hash, Db>
 where
-	B::Hash: Unpin,
-	NumberFor<B>: Into<u32>,
+	Block: BlockT + Unpin,
+	Db: ReadOnlyDb + 'static,
+	Block::Hash: Unpin,
+	NumberFor<Block>: Into<u32>,
 {
-	storage: Address<workers::StorageAggregator<B>>,
-	blocks: Address<workers::BlocksIndexer<B, D>>,
-	metadata: Address<workers::MetadataActor<B>>,
-	db_pool: Address<ActorPool<DatabaseActor<B>>>,
+	async fn spawn(conf: &SystemConfig<Block, Db>) -> Result<Self> {
+		let mut tasks = Vec::new();
+		let db = workers::DatabaseActor::new(conf.pg_url().into(), conf.executor.clone()).await?.create(None);
+		let (db, fut) = db.run();
+		tasks.push(conf.executor.spawn(fut));
+		let storage = workers::StorageAggregator::new(db.clone()).create(None);
+		let metadata = workers::MetadataActor::new(db.clone(), conf.meta().clone()).await?.create(None);
+		let (storage, fut) = storage.run();
+		tasks.push(conf.executor.spawn(fut));
+		let (metadata, fut) = metadata.run();
+		tasks.push(conf.executor.spawn(fut));
+		let blocks = workers::BlocksIndexer::new(conf, db.clone(), metadata.clone()).create(None);
+		let (blocks, fut) = blocks.run();
+		tasks.push(conf.executor.spawn(fut));
+
+		Ok(Actors { storage, blocks, metadata, db, tasks })
+	}
+
+	// Run a future that sends actors a signal to progress every X seconds
+	async fn tick_interval(&self, executor: Arc<smol::Executor<'_>>) -> Result<()> {
```

Comment:
Not sure I see where the `x seconds` part is happening, can you point it out for me?